<div class="alert alert-block alert-danger">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

# Comentario General

Hola, Alejandro. Me presento: soy David Bautista, el revisor humano que te ayudará en este proceso tras la guía inicial de Dot.

He revisado tu avance y he dejado comentarios específicos en las secciones correspondientes para que los puedas tener en cuenta de cara a tu siguiente entrega. 

¡Mucho éxito con los ajustes y quedo atento a tus dudas!
</div>

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

# Comentario General #2

Hola, Alejandro. Gracias por tener en cuenta los comentarios. Muy buen trabajo con las modificaciones. Te felicito por la culminación del proyecto.
</div>


# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [2]:
# Importa pandas
import pandas as pd

<div class="alert alert-block alert-success">
<b>Comentario de Revisor #2 </b> <a class="tocSkip"></a>

Perfecto, buen trabajo importando ``pandas``.  </div>


<details>
<summary>Haz clic para ver la pista</summary>

Para usar `pandas`, necesitas la palabra clave `import` seguida del nombre de la biblioteca y un alias.


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [3]:
# Lee el archivo y almacénalo en df
df=pd.read_csv('/datasets/music_project_en.csv')

<div class="alert alert-block alert-success">
<b>Comentario de Revisor #2</b> <a class="tocSkip"></a>

Excelente, buen trabajo cargando el dataset.  </div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa la palabra clave `read_csv` para leer el archivo CSV. Asegúrate de incluir la ruta `/datasets/music_project_en.csv` y guarda el resultado en la variable `df`.

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [4]:
# Obtén las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<div class="alert alert-block alert-success">
<b>Comentario de Revisor #2</b> <a class="tocSkip"></a>

Perfecto, buen trabajo usando el método ``head()``.  </div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `head` para ver las primeras filas del DataFrame.

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [5]:
# Obtén la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #2</b> <a class="tocSkip"></a>

Excelente, buen trabajo usando el método ``info()``.  </div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `info` para obtener detalles sobre columnas, tipos de datos y valores nulos.


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos nombres de las columnas.




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

# Escribe aquí tus respuestas:

1.El tipo de dato presente en las filas es tipo object. En el diccionario de datos se encuentra el nombre de cada columna y que almacena.

2.Aun no hay suficientes datos para probar la hipótesis, ya que aun falta limpiar los datos, procesarlos e interpretar dichos datos. Lo que se puede hacer con la información actual es percatarse de posibles problemas de datos nulos, el tipo de los datos y las columnas.

3.En la tabla con esta información se observo que hay valores ausentes, esto se presento en las siguientes columnas:Track,artist y genre.

<div class="alert alert-block alert-success">
<b>Comentario de Revisor #2</b> <a class="tocSkip"></a>

Buen trbajo con el desarrollo de las observaciones.  </div>


## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [6]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #2</b> <a class="tocSkip"></a>

Perfecto, buen trabajo mostrando los encabezados del dataset.   </div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el atributo `.columns` para ver todos los nombres de las columnas en el DataFrame.


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [7]:
# Bucle que itera sobre los encabezados y los pone todos en minúsculas
list=[]
for i in df.columns:
    list.append(i.lower())

df.columns=list
print(df.columns)

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa un `for` para recorrer `df.columns` y aplica `.lower()` a cada nombre. Luego, asigna la lista resultante de nuevo a `df.columns`.


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [8]:
# Bucle que itera sobre los encabezados y elimina los espacios
list=[]
for i in df.columns:
    list.append(i.strip())

df.columns=list
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.strip()` dentro del bucle para limpiar los nombres de columnas. Asegúrate de reasignar los nuevos nombres a `df.columns`.


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [9]:
# Cambia el nombre de la columna "userid"
df.rename(columns={'userid':'user_id'},inplace=True)
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id`. No olvides establecer `inplace=True` o reasignar a `df`.

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [10]:
# Comprueba el resultado: lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor #2</b> <a class="tocSkip"></a>

Excelente, buen trabajo modificando los encabezados.  </div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [11]:
# Calcula el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo identificando la cantidad de valores nulos presentes en cada columna del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.isna()` para detectar los valores ausentes y combínalo con el método `.sum()` para contar cuántos hay por columna.


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [12]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace=['track','artist','genre']
for i in columns_to_replace:
    df[i].fillna('unknown',inplace=True)

<details>
<summary>Haz clic para ver la pista</summary>

Crea una lista con los nombres de las columnas y usa un `for` para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [13]:
# Cuenta los valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Excelente, buen trabajo usando el bucle con el método <code>fillna()</code> para solucionar el problema de los valores nulos dentro del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `isna()` con `sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [14]:
# Cuenta los duplicados explícitos
print(df.duplicated().sum())

3826


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo identificando la cantidad de valores duplicados explicitos dentro del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa un método que detecte filas repetidas (`duplicated`) y luego otro para contarlas.

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [15]:
# Elimina los duplicados explícitos
df.drop_duplicates(inplace=True)

<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo usando el método <code>drop_duplicates()</code> para solucionar el problema de los valores duplicados en el dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Busca un método que elimine las filas duplicadas directamente del DataFrame. No olvides actualizar la variable.

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [16]:
# Comprueba de nuevo si hay duplicados
print(df.duplicated().sum())

0


<details>
<summary>Haz clic para ver la pista</summary>

Repite el mismo enfoque que usaste para contar duplicados antes. Deberías ver que el total ahora es cero.

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [25]:
# Inspecciona los nombres de géneros únicos
print(sorted(df['genre'].unique()))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

<details>
<summary>Haz clic para ver la pista</summary>

Usa `unique()` para obtener los géneros únicos. Si quieres verlos en orden alfabético, combina con `sorted()`.

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.


1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.


In [21]:

# Función para reemplazar los duplicados implícitos
def replace_wrong_values(df,column,wrong_values,correct_value):
    for i in wrong_values:
        df[column].replace(i,correct_value,inplace=True)

       
            
        
        

            

<details>
<summary>Haz clic para ver la pista</summary>

Crea una función con cuatro parámetros. Dentro del bucle, llama a `.replace()` sobre la columna para cada valor incorrecto, cambiándolo por el correcto.

Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [22]:
# Elimina los duplicados implícitos
replace_wrong_values(df,'genre',['hip','hop','hip-hop'],'hiphop')

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Excelente, buen trabajo estructurando y aplicando la función para solucionar el problema de los valores duplicados implícitos en la columna con información de género.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Llama a la función que creaste y pasa los cuatro argumentos necesarios. Asegúrate de usar el nombre correcto de la columna y lista los valores que deben unificarse.

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [26]:
# Comprueba de nuevo los duplicados implícitos
print(sorted(df['genre'].unique()))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

## Etapa 3. Análisis

<details>
<summary>Haz clic para ver la pista</summary>

Usa de nuevo el método que muestra los valores únicos de una columna para verificar que todos los nombres se hayan unificado correctamente.

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [46]:
# Cuenta las canciones reproducidas en cada ciudad
df_filtered = df.query('(city=="Springfield" or city=="Shelbyville") and (day=="Monday" or day=="Friday")')
conteo_ciudad = df_filtered.groupby('city')['track'].count()
generos_unicos = df_filtered.groupby('city')['genre'].nunique()
print('Reproducciones por ciudad:')
print(conteo_ciudad)
print('Generos unicos por ciudad')
print(generos_unicos)

Reproducciones por ciudad:
city
Shelbyville    11509
Springfield    31685
Name: track, dtype: int64
Generos unicos por ciudad
city
Shelbyville    181
Springfield    240
Name: genre, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Agrupa los datos por ciudad con el método `groupby()` y usa el método `count()` para contar las canciones reproducidas en cada grupo.

Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?

Escribe tus observaciones aquí.

En las tablas se pueden observar tanto diferencias como similitudes entre ambas ciudades en los días lunes y viernes. Dentro de las diferencias principales se encuentra una diferencia significativa entre el total de canciones reproducidas, ya que shelbyville cuenta con 11509, mientras que springfield cuenta con un total de 31685. En adición a ello se logra observar una disparidad en cuando a diversidad de géneros, ya que en springfield se escucharon 240 géneros únicos y shelbyville escucho 181 géneros diferentes. Por consiguiente se puede afirmar que hay una correlación entre canciones escuchadas y diversidad de géneros. Dicha disparidad puede ser causada por muchos factores demográficos, entre ellos el mas determinante siendo la población total.

Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.



In [47]:
# Calcula las canciones reproducidas en cada uno de los dos días
df_filtered = df.query('(city=="Springfield" or city=="Shelbyville") and (day=="Monday" or day=="Friday")')
conteo_dia = df_filtered.groupby(['city', 'day'])['track'].count()
print(conteo_dia)


city         day   
Shelbyville  Friday     5895
             Monday     5614
Springfield  Friday    15945
             Monday    15740
Name: track, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Cambia el criterio de agrupación a la columna que representa el día de la semana y filtra solo para los días que te interesan.

Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

Escribe tus observaciones aquí.

Tanto el lunes como el viernes mantuvieron una actividad similar con una diferencia minima en ambas ciudades. Si se analiza el comportamiento de las ciudades individualmente por día se puede observar que en springfield escuchan mas canciones que en shelvibyville los lunes y los viernes. además se puede observar en ambos casos que el día presenta una tendencia a mayor actividad de usuarios respecto a los lunes en ambas ciudades.


Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [48]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day,city):
    df_new=df.query('day==@day and city==@city')
    return df_new['user_id'].count()
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=

    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=

    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()

    # Devuelve el número de valores de la columna 'user_id'

<details>
<summary>Haz clic para ver la pista</summary>

Define una función con dos parámetros. Dentro, aplica un filtrado secuencial usando `df[...]`, y al final cuenta los valores con `.count()`.

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [49]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks('Monday','Springfield')


15740

In [50]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks('Monday','Shelbyville')


5614

In [51]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks('Friday','Springfield')


15945

In [52]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks('Friday','Shelbyville')


5895

In [53]:
print('Diferencia promedio por ciudad')
print(((15740+15945)/2)-((5614+5895)/2))



Diferencia promedio por ciudad
10088.0


In [54]:
print("Diferencia promedio por dia")
print(((15945-15740)+(5895-5614))/2)

Diferencia promedio por dia
243.0


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo estructurando y aplicando la función.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Llama a tu función con combinaciones distintas de ciudad y día, como `'Monday', 'Springfield'`, y guarda o imprime los resultados para comparar.

# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


Detalla aquí tus conclusiones.

Durante el proceso de análisis se identificaron patrones claros en el comportamiento de los usuarios en relación con la música que escuchan, siendo los principales la ciudad y el día de la semana. El factor más influyente fue la ciudad, ya que en Springfield se registraron 15 740 usuarios los lunes y 15 945 los viernes, mientras que en Shelbyville los valores fueron considerablemente menores, con 5 614 usuarios los lunes y 5 895 los viernes. Esta diferencia generó una disparidad significativa en la diversidad de géneros musicales, pues en Springfield se escucharon 240 géneros únicos, frente a 181 en Shelbyville. Esto evidencia una mayor concentración de oyentes en Springfield, con una diferencia media de 10 088 usuarios entre ciudades y, en consecuencia, una mayor diversidad musical. En cuanto al día de la semana, se observa una influencia menor, ya que los viernes presentan un número ligeramente superior de usuarios respecto a los lunes, con una diferencia promedio de 243 usuarios. En conjunto, estos resultados indican una relación significativa entre la ciudad, el día y la música que escuchan los usuarios.

Durante el proceso de análisis se identificaron diversos problemas relacionados con la calidad de los datos. Inicialmente, el conjunto de datos presentaba columnas con nombres no estandarizados, valores ausentes, inconsistencias en los nombres de los géneros musicales y registros duplicados. Estas dificultades fueron solucionadas mediante un proceso de limpieza y estandarización de los datos, lo que permitió un tratamiento más adecuado de la información.

Estas acciones contribuyeron de manera significativa a mejorar la calidad del análisis, ya que, sin ellas, los resultados habrían estado sesgados por la presencia de datos incompletos y duplicados, afectando la exactitud de las conclusiones. Por lo tanto, la limpieza y organización de los datos fue un paso fundamental para garantizar un análisis confiable y representativo del comportamiento real de los usuarios.

<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Buen trabajo. 
</div>
